contracts/TwabRewards.sol

```diff
@@ -0,0 +1,392 @@
+        address _user,
+        uint256 _promotionId,
+        uint256[] calldata _epochIds
+    ) external override returns (uint256) {
+        uint256 _rewardAmount;
+
+        uint256 _userClaimedEpochs = _claimedEpochs[_promotionId][_user];
+
+        for (uint256 index = 0; index < _epochIds.length; index++) {
+            uint256 _epochId = _epochIds[index];
+
+            require(_epochId < _getCurrentEpochId(_promotionId), "TwabRewards/epoch-not-over");
+
+            require(
+                !_isClaimedEpoch(_user, _promotionId, _epochId),
+                "TwabRewards/rewards-already-claimed"
+            );
+
+            _rewardAmount += _calculateRewardAmount(_user, _promotionId, _epochId);
+            _userClaimedEpochs = _updateClaimedEpoch(_userClaimedEpochs, _epochId);
+        }
+
+        _claimedEpochs[_promotionId][_user] = _userClaimedEpochs;
+
+        IERC20 _token = IERC20(_getPromotion(_promotionId).token);
+        _token.safeTransferFrom(address(this), _user, _rewardAmount);
+
+        emit RewardsClaimed(_promotionId, _epochIds, _rewardAmount);
+
+        return _rewardAmount;
+    }
+
+    /* ============ Internal Functions ============ */
+
+    /**
+    @notice Determine if address passed is actually a ticket.
+    @param _ticket Address to check
+   */
+    function _requireTicket(address _ticket) internal view {
+        require(_ticket != address(0), "TwabRewards/ticket-not-zero-address");
+
+        (bool succeeded, bytes memory data) = address(_ticket).staticcall(
+            abi.encodePacked(ITicket(_ticket).controller.selector)
+        );
+
+        address controllerAddress;
+
+        if (data.length > 0) {
+            controllerAddress = abi.decode(data, (address));
+        }
+
+        require(succeeded && controllerAddress != address(0), "TwabRewards/invalid-ticket");
+    }
+
+    /**
+        @notice Get settings for a specific promotion.
+        @dev Will revert if the promotion does not exist.
+        @param _promotionId Promotion id to get settings for
+        @return Promotion settings
+     */
+    function _getPromotion(uint256 _promotionId) internal view returns (Promotion memory) {
+        return _promotions[_promotionId];
+    }
+
+    /**
+        @notice Get the current epoch id of a promotion.
+        @dev Epoch ids and their boolean values are tightly packed and stored in a uint256, so epoch id starts at 0.
+        @param _promotionId Id of the promotion to get current epoch for
+        @return Epoch id
+     */
+    function _getCurrentEpochId(uint256 _promotionId) internal view returns (uint256) {
+        Promotion memory _promotion = _getPromotion(_promotionId);
+        uint256 _numberOfEpochs = _promotion.numberOfEpochs;
+
+        // (_numberOfEpochs * elapsedTimestamp) / promotionDurationTimestamp
+        return
+            (_numberOfEpochs * (block.timestamp - _promotion.startTimestamp)) /
+            (_promotion.epochDuration * _numberOfEpochs);
```

Comment:
?

why are you multiplying both the numerator and denominator by the _numberOfEpochs ?

Reply:
Cause I multiply the number of epochs by the elapsed time and then divide by the total duration of the promotion to get the current epoch id. But you're right, I can divide the elapsed timestamp by the epoch duration timestamp.
Fixed in this commit: https://github.com/pooltogether/v4-periphery/pull/8/commits/21ec50cc74d72fa9cd1f18a8f2908e700ecf405e